Objects/frameobject.c

```diff
@@ -13,9 +13,320 @@
+    PyCodeObject *co = _PyFrame_GetCode(frame);
+
+    for (int i = 0; i < co->co_nlocalsplus; i++) {
+        if (_PyUnicode_EQ(PyTuple_GET_ITEM(co->co_localsplusnames, i), key)) {
+            _PyLocals_Kind kind = _PyLocals_GetKind(co->co_localspluskinds, i);
+
+            PyObject *oldvalue = fast[i];
+            PyObject *cell = NULL;
+            if (kind == CO_FAST_FREE) {
+                // The cell was set when the frame was created from
+                // the function's closure.
+                assert(oldvalue != NULL && PyCell_Check(oldvalue));
+                cell = oldvalue;
+            } else if (kind & CO_FAST_CELL && oldvalue != NULL) {
+                /* Same test as in PyFrame_FastToLocals() above. */
+                if (PyCell_Check(oldvalue)) {
+                    cell = oldvalue;
+                }
+            }
+            if (cell != NULL) {
+                oldvalue = PyCell_GET(cell);
+                if (value != oldvalue) {
+                    PyCell_SET(cell, Py_XNewRef(value));
+                    Py_XDECREF(oldvalue);
+                }
+            } else if (value != oldvalue) {
+                if (value == NULL) {
+                    // Probably can't delete this, since the compiler's flow
+                    // analysis may have already "proven" that it exists here:
+                    const char *e = "assigning None to unbound local %R";
+                    if (PyErr_WarnFormat(PyExc_RuntimeWarning, 0, e, key)) {
+                        // It's okay if frame_obj is NULL, just try anyways:
+                        PyErr_WriteUnraisable((PyObject *)frame->frame_obj);
+                    }
+                    value = Py_NewRef(Py_None);
+                }
+                Py_XSETREF(fast[i], Py_NewRef(value));
+            }
+            Py_XDECREF(value);
+            return 0;
+        }
+    }
+
+    // Okay not in the fast locals, try extra locals
+
+    PyObject *extra = frame->f_extra_locals;
+
+    if (extra != NULL) {
+        assert(PyDict_Check(extra));
+        if (value == NULL) {
+            if (PyDict_DelItem(extra, key) < 0) {
+                return -1;
+            }
+        } else {
+            if (PyDict_SetItem(extra, key, value) < 0) {
+                return -1;
+            }
+        }
+    }
+
+    return 0;
+}
+
+static PyMappingMethods framelocalsproxy_as_mapping = {
+    framelocalsproxy_length, // mp_length
+    framelocalsproxy_getitem, // mp_subscript
+    framelocalsproxy_setitem, // mp_ass_subscript
+};
+
+static PyMethodDef framelocalsproxy_methods[] = {
+    {"__getitem__",     framelocalsproxy_getitem,                 METH_O | METH_COEXIST,
+     NULL},
+    {"keys",     framelocalsproxy_keys,                 METH_NOARGS,
+     NULL},
+    {"items",     framelocalsproxy_items,                 METH_NOARGS,
+     NULL},
+    {NULL,              NULL}   /* sentinel */
+};
+
+PyTypeObject PyFrameLocalsProxy_Type = {
```

Comment:
Can you use C named initializers, please. https://en.cppreference.com/w/c/language/struct_initialization
E.g.
```
    .tp_dealloc = &framelocalsproxy_dealloc,
```
it saves a lot of boilerplate and is less error prone.